File: pipeline/plan.py
"""What a configuration will actually do, before it does it.

The run form's estimate was one line from one measured constant. It did not say how
many image-visits the configuration implies, that a centralised ceiling needs the
same number to be worth comparing, which stages will really run, or what the
equivalent command is.

That arithmetic being in somebody's head rather than on screen is exactly how a
baseline shipped this morning with 1.667x the federation's budget.

    python -m pipeline.plan --profile full --vehicles 6 --rounds 6 --epochs 4
"""
from __future__ import annotations

import argparse

from . import holdout, stages, vehicles
from .stages import Config

#: Measured on an RTX 5070 Ti, 6 vehicles x 1 400 images x 6 rounds x 4 epochs:
#: 3 296 s and 82.2 Wh for 201 600 image-visits at 640 px.
SECONDS_PER_KVISIT = 3296 / 201.6
WH_PER_KVISIT = 82.2 / 201.6


def budget(cfg: Config) -> dict:
    """The arithmetic both sides of the comparison have to agree on."""
    visits = cfg.n_vehicles * cfg.per_vehicle * cfg.rounds * cfg.local_epochs
    pooled = cfg.n_vehicles * cfg.per_vehicle
    # Scaled by resolution: cost goes with pixels, and demo runs at 320 not 640.
    scale = (cfg.imgsz / 640) ** 2
    return {
        "vehicles": cfg.n_vehicles,
        "images_per_vehicle": cfg.per_vehicle,
        "rounds": cfg.rounds,
        "local_epochs": cfg.local_epochs,
        "image_visits": visits,
        "effective_epochs": cfg.rounds * cfg.local_epochs,
        "pooled_images": pooled,
        "centralised_epochs_to_match": cfg.rounds * cfg.local_epochs,
        "seconds_estimate": round(visits / 1000 * SECONDS_PER_KVISIT * scale),
        "wh_estimate": round(visits / 1000 * WH_PER_KVISIT * scale, 1),
        "imgsz": cfg.imgsz,
    }


def commands(cfg: Config) -> list[dict]:
    """The exact CLI for this configuration, and the comparisons it enables."""
    base = (f"python -m pipeline.runner --all --profile {cfg.profile} "
            f"--vehicles {cfg.n_vehicles} --rounds {cfg.rounds} --epochs {cfg.local_epochs} "
            f"--seed {cfg.seed} --partition {cfg.partition} --strategy {cfg.strategy}")
    if cfg.per_vehicle_override:
        base += f" --per-vehicle {cfg.per_vehicle_override}"
    if cfg.partition == "dirichlet":
        base += f" --alpha {cfg.alpha}"
    per_vehicle = f" --per-vehicle {cfg.per_vehicle_override}" if cfg.per_vehicle_override else ""
    sweep = (f"--profile {cfg.profile} --vehicles {cfg.n_vehicles} --rounds {cfg.rounds} "
             f"--epochs {cfg.local_epochs}{per_vehicle} --yes")
    return [
        {"label": "this run", "cmd": base + " --yes --skip baseline"},
        {"label": "this run, with the matched centralised ceiling", "cmd": base + " --yes"},
        {"label": "is it just the seed?", "cmd": f"python -m pipeline.experiment --preset seeds --seeds 0,1,2 {sweep}"},
        {"label": "does the strategy matter?", "cmd": f"python -m pipeline.experiment --preset strategies --strategies fedavg,fedadam,fedavgm {sweep}"},
        {"label": "does non-IID matter?", "cmd": f"python -m pipeline.experiment --preset partitions --partitions condition,random,dirichlet {sweep}"},
        {"label": "how much does skew move it?", "cmd": f"python -m pipeline.experiment --preset alpha --alphas 0.05,0.5,100 {sweep}"},
        {"label": "compare what you already have", "cmd": "python -m pipeline.compare --last 10 --md"},
    ]


def warnings(cfg: Config) -> list[str]:
    """What this configuration will do that the person may not have intended."""
    out = []
    if not holdout.names():
        out.append("No holdout carved. Every number this run produces will be a client "
                   "scoring itself on its own distribution, and comparable with nothing.")
    if cfg.partition == "condition":
        # BDD's rarest profiled condition; asking for more silently tops up.
        thinnest = 1419
        if cfg.per_vehicle > thinnest:
            out.append(f"{cfg.per_vehicle} images per vehicle exceeds the rarest condition "
                       f"(~{thinnest} images in all of BDD100K), so those shards will top up "
                       f"with unrelated images and the run will be closer to IID than it looks.")
    if cfg.rounds < 2:
        out.append("One round cannot show whether the global model moves, which is the "
                   "single signal that says the federation is learning at all.")
    if cfg.n_vehicles > 8:
        out.append("More vehicles than condition profiles; the extras repeat conditions.")
    b = budget(cfg)
    if b["seconds_estimate"] > 3600:
        out.append(f"About {b['seconds_estimate'] // 60} minutes of GPU time. Vehicles train "
                   f"serialised, so wall clock scales with vehicle count.")
    return out


def plan(cfg: Config) -> dict:
    return {
        "config": cfg.to_dict(),
        "budget": budget(cfg),
        "stages": stages.snapshot(cfg),
        "commands": commands(cfg),
        "warnings": warnings(cfg),
        "fleet_on_disk": vehicles.load_fleet_meta(),
        "holdout": holdout.meta(),
    }


def main(argv=None) -> int:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--profile", default="demo", choices=("demo", "full"))
    ap.add_argument("--vehicles", type=int, default=6)
    ap.add_argument("--rounds", type=int, default=2)
    ap.add_argument("--epochs", type=int, default=1)
    ap.add_argument("--per-vehicle", type=int, default=0)
    ap.add_argument("--partition", default="condition", choices=vehicles.PARTITIONS)
    ap.add_argument("--strategy", default="fedavg", choices=stages.STRATEGIES)
    ap.add_argument("--seed", type=int, default=0)
    args = ap.parse_args(argv)

    cfg = Config(profile=args.profile, n_vehicles=args.vehicles, rounds=args.rounds,
                 local_epochs=args.epochs, per_vehicle_override=args.per_vehicle,
                 partition=args.partition, strategy=args.strategy, seed=args.seed)
    p = plan(cfg)
    b = p["budget"]

    print(f"{b['vehicles']} vehicles x {b['images_per_vehicle']} images x {b['rounds']} rounds "
          f"x {b['local_epochs']} local epochs")
    print(f"  = {b['image_visits']:,} image-visits at {b['imgsz']}px "
          f"({b['effective_epochs']} effective epochs per vehicle)")
    print(f"  ~ {b['seconds_estimate'] // 60} min, ~{b['wh_estimate']} Wh")
    print(f"  a matched centralised ceiling: {b['pooled_images']:,} pooled images x "
          f"{b['centralised_epochs_to_match']} epochs\n")

    print("stages:")
    for s in p["stages"]:
        mark = "skip" if s["satisfied"] else ("GATE" if s["gated"] else "run ")
        print(f"  [{mark}] {s['name']:<9} {s['est']:<18} {s['detail'][:60]}")

    if p["warnings"]:
        print("\nwarnings:")
        for w in p["warnings"]:
            print(f"  ! {w}")

    print("\ncommands:")
    for c in p["commands"]:
        print(f"  # {c['label']}\n  {c['cmd']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: pipeline/static/js/plan.js
// Plan view: what the configuration on the Control tab will actually do.
//
// The estimate used to be one line from one measured constant. It did not say how
// many image-visits the configuration implies, that a centralised ceiling needs the
// same number to be comparable, or what the equivalent command is. That arithmetic
// living in somebody's head is how a baseline shipped with 1.667x the budget.
import { $, esc, empty } from "./util.js";

let current = null;

export async function loadPlan() {
  $("planBody").innerHTML = '<div class="skel-block" style="min-height:240px"></div>';
  try {
    current = await (await fetch("/api/plan")).json();
  } catch {
    $("planBody").innerHTML = empty("The server did not answer.", "Is it still running?");
    return;
  }
  render();
}

const mins = (s) => s < 90 ? `${s} s` : `${Math.round(s / 60)} min`;

function render() {
  const p = current;
  if (!p) return;
  const b = p.budget, c = p.config;

  $("planBody").innerHTML =
    `<div class="panel"><h2>Budget<span class="n">what both sides of a comparison must agree on</span></h2>` +
      `<div class="sum">${b.vehicles} vehicles <em>×</em> ${b.images_per_vehicle.toLocaleString()} images ` +
      `<em>×</em> ${b.rounds} rounds <em>×</em> ${b.local_epochs} local epochs ` +
      `<em>=</em> <b>${b.image_visits.toLocaleString()}</b> image-visits</div>` +
      `<div class="readouts" style="margin-top:var(--s4)">` +
        `<div class="readout"><span class="v">${b.effective_epochs}</span>` +
          `<div class="k">effective epochs per vehicle</div></div>` +
        `<div class="readout"><span class="v">${mins(b.seconds_estimate)}</span>` +
          `<div class="k">estimated GPU time</div></div>` +
        `<div class="readout"><span class="v">${b.wh_estimate}</span><span class="u">Wh</span>` +
          `<div class="k">estimated energy</div></div>` +
        `<div class="readout"><span class="v">${b.imgsz}</span><span class="u">px</span>` +
          `<div class="k">image size</div></div>` +
        `<div class="readout"><span class="v">${b.pooled_images.toLocaleString()}</span>` +
          `<div class="k">pooled images for the ceiling</div></div>` +
        `<div class="readout"><span class="v">${b.centralised_epochs_to_match}</span>` +
          `<div class="k">epochs to match it</div></div>` +
      `</div>` +
      `<p class="hint">A centralised model trained on those ${b.pooled_images.toLocaleString()} ` +
      `images for ${b.centralised_epochs_to_match} epochs makes exactly the same ` +
      `${b.image_visits.toLocaleString()} image-visits. Give it more and the gap measures ` +
      `the budget rather than the method — which is what happened the first time this ` +
      `project ran one.</p>` +
      `<p class="hint">Estimates are calibrated on a measured run: 6 vehicles × 1 400 ` +
      `images × 6 rounds × 4 epochs took 3 296 s and 82.2 Wh on an RTX 5070 Ti.</p></div>` +

    (p.warnings && p.warnings.length
      ? `<div class="panel"><h2>Before you start</h2><ul class="plain">` +
        p.warnings.map(w => `<li><span class="lamp s-needs_confirm">check</span>` +
          `<span style="color:var(--ink-2)">${esc(w)}</span></li>`).join("") + `</ul></div>`
      : "") +

    `<div class="panel"><h2>Stages<span class="n">for this configuration</span></h2>` +
      `<table><thead><tr><th scope="col">Stage</th><th scope="col">State</th>` +
      `<th scope="col">Cost</th><th scope="col">Why</th></tr></thead><tbody>` +
      p.stages.map(s => {
        const cls = s.satisfied ? "s-skipped" : (s.gated ? "s-needs_confirm" : "s-pending");
        const word = s.satisfied ? "skip" : (s.gated ? "gated" : "will run");
        return `<tr><td><b>${esc(s.name)}</b><br><span style="color:var(--dim);` +
          `font-size:var(--t-sm)">${esc(s.title)}</span></td>` +
          `<td><span class="lamp ${cls}">${word}</span></td>` +
          `<td class="num">${esc(s.est)}</td>` +
          `<td style="color:var(--dim);font-size:var(--t-sm)">${esc(s.detail)}</td></tr>`;
      }).join("") + `</tbody></table></div>` +

    `<div class="panel"><h2>The same thing, from a terminal` +
      `<span class="n">no dashboard required</span></h2>` +
      p.commands.map((cmd, i) =>
        `<div class="cmd"><div class="lbl">${esc(cmd.label)}</div>` +
        `<pre class="cmdline" id="cmd${i}">${esc(cmd.cmd)}</pre>` +
        `<button class="copy" data-cmd="${i}">Copy</button></div>`).join("") +
      `<p class="hint">Every arm of a comparison changes exactly one setting and ends ` +
      `scored on the same holdout, which is the only number comparable between runs.</p></div>` +

    `<div class="panel"><h2>On disk right now</h2><dl class="kvs">` +
      `<dt>fleet partition</dt><dd>${esc((p.fleet_on_disk || {}).partition ?? "none")}</dd>` +
      `<dt>fleet seed</dt><dd>${esc(String((p.fleet_on_disk || {}).seed ?? "—"))}</dd>` +
      `<dt>images per vehicle</dt><dd>${esc(String((p.fleet_on_disk || {}).per_vehicle ?? "—"))}</dd>` +
      `<dt>fleet fingerprint</dt><dd>${esc((p.fleet_on_disk || {}).fingerprint ?? "—")}</dd>` +
      `<dt>holdout</dt><dd>${esc(String((p.holdout || {}).size ?? "none"))} images, seed ` +
      `${esc(String((p.holdout || {}).seed ?? "—"))}</dd>` +
      `<dt>strategy this run</dt><dd>${esc(c.strategy)}</dd>` +
      `<dt>partition this run</dt><dd>${esc(c.partition)}` +
      `${c.partition === "dirichlet" ? ` α=${c.alpha}` : ""}</dd>` +
      `</dl><p class="hint">If the fleet on disk does not match the configuration, the ` +
      `fleet stage rebuilds it rather than reusing shards that answer a different ` +
      `question.</p></div>`;

  document.querySelectorAll("#planBody .copy").forEach(b => b.onclick = async () => {
    const text = $("cmd" + b.dataset.cmd).textContent;
    try {
      await navigator.clipboard.writeText(text);
      b.textContent = "Copied";
    } catch {
      // Clipboard is blocked on http:// in some browsers; select it instead so the
      // keyboard can still take it.
      const range = document.createRange();
      range.selectNodeContents($("cmd" + b.dataset.cmd));
      const sel = window.getSelection();
      sel.removeAllRanges();
      sel.addRange(range);
      b.textContent = "Selected";
    }
    setTimeout(() => { b.textContent = "Copy"; }, 1600);
  });
}
